src/btree/bt_sync_obsolete.c

```diff
@@ -291,3 +290,324 @@ __wt_sync_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *parent)
+    conn = S2C(session);
+
+    __wt_seconds(session, &last);
+    for (;;) {
+        /* Check periodically in case the signal was missed. */
+        __wt_cond_wait_signal(session, conn->cc_cleanup.cond, 5 * WT_MILLION,
+          __checkpoint_cleanup_run_chk, &cv_signalled);
+
+        /* Check if we're quitting. */
+        if (!__checkpoint_cleanup_run_chk(session))
+            break;
+
+        __wt_seconds(session, &now);
+
+        /*
+         * See if it is time to checkpoint cleanup. Checkpoint cleanup is an operation that
+         * typically has long intervals so skipping some should have little impact.
+         */
+        if (!cv_signalled && (now - last < 60))
+            continue;
+
+        WT_ERR(__checkpoint_cleanup_int(session));
+        WT_STAT_CONN_INCR(session, checkpoint_cleanup_success);
+        last = now;
+    }
+
+err:
+    if (ret != 0)
+        WT_IGNORE_RET(__wt_panic(session, ret, "checkpoint cleanup error"));
+    return (WT_THREAD_RET_VALUE);
+}
+
+/*
+ * __wt_checkpoint_cleanup_create --
+ *     Start the checkpoint cleanup thread.
+ */
+int
+__wt_checkpoint_cleanup_create(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    uint32_t session_flags;
+
+    conn = S2C(session);
+
+    if (F_ISSET(conn, WT_CONN_IN_MEMORY | WT_CONN_READONLY))
+        return (0);
+
+    /* Set first, the thread might run before we finish up. */
+    FLD_SET(conn->server_flags, WT_CONN_SERVER_CHECKPOINT_CLENAUP);
+
+    /*
+     * Checkpoint cleanup does enough I/O it may be called upon to perform slow operations for the
+     * block manager.
+     */
+    session_flags = WT_SESSION_CAN_WAIT;
+    WT_RET(__wt_open_internal_session(
+      conn, "checkpoint-cleanup", true, session_flags, 0, &conn->cc_cleanup.session));
+    session = conn->cc_cleanup.session;
+
+    WT_RET(__wt_cond_alloc(session, "checkpoint cleanup", &conn->cc_cleanup.cond));
+
+    WT_RET(__wt_thread_create(session, &conn->cc_cleanup.tid, __checkpoint_cleanup, session));
+    conn->cc_cleanup.tid_set = true;
+
+    return (0);
+}
+
+/*
+ * __wt_checkpoint_cleanup_destroy --
+ *     Destroy the checkpoint cleanup thread.
+ */
+int
+__wt_checkpoint_cleanup_destroy(WT_SESSION_IMPL *session)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+
+    conn = S2C(session);
+
+    FLD_CLR(conn->server_flags, WT_CONN_SERVER_CHECKPOINT_CLENAUP);
+    if (conn->cc_cleanup.tid_set) {
+        __wt_cond_signal(session, conn->cc_cleanup.cond);
+        WT_TRET(__wt_thread_join(session, &conn->cc_cleanup.tid));
+        conn->cc_cleanup.tid_set = false;
+    }
+    __wt_cond_destroy(session, &conn->cc_cleanup.cond);
+
+    /* Close the server thread's session. */
+    if (conn->cc_cleanup.session != NULL) {
+        WT_TRET(__wt_session_close_internal(conn->cc_cleanup.session));
+        conn->cc_cleanup.session = NULL;
+    }
+
+    return (ret);
+}
+
+/*
+ * __wt_checkpoint_cleanup_trigger --
+ *     Trigger the checkpoint cleanup thread.
+ */
+void
```

Comment:
Is this also a debug function? May be add debug into the function name?

Reply:
This function is used for triggering the checkpoint cleanup from the checkpoint. I am thinking of this function use case from the `compact` as the checkpoint doesn't clean the obsolete content now, we may need to trigger the checkpoint cleanup also. I want to leave it as is for now.